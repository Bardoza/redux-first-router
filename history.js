module.exports = {
  createHistory: require('./dist/core/createHistory.js').default,
  History: require('./dist/history/History.js').default,
  MemoryHistory: require('./core/history/MemoryHistory.js').default,
  BrowserHistory: require('./core/history/BrowserHistory.js').default,
  utils: {
    dom: require('./dist/history/utils/dom.js'),
    location: require('./core/history/utils/location.js'),
    path: require('./dist/history/utils/path.js'),
    sessionStorage: require('./core/history/utils/sessionStorage.js'),
  }
}


/** if you want to extend History, here is how you do it:

import History from 'rudy/history/BrowserHistory'

class MyHistory extends History {
  push(path) {
    const location = this.createAction(path)
    // do something custom
  }
}

// usage:

import { createRouter } from 'rudy'
import { createHistory as creatHist } from 'rudy/core'

const createHistory = (routes, opts) => {
  if (opts.someCondition) return new MyHistory(routes, opts)
  return creatHist(routes, opts)
}

const { middleware, reducer, firstRoute } = createRouter(routes, { createHistory })

*/
