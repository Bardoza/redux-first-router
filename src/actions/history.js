// @flow
import { CALL_HISTORY } from '../types'

export const push = (path: string, state: ?Object) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'push',
    args: [path, state]
  }
})

export const replace = (path: string, state: ?Object) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'replace',
    args: [path, state]
  }
})

export const jump = (n: number | string, state: ?Object, byIndex: ?boolean, kind: ?string) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'jump',
    args: [n, state, byIndex, kind]
  }
})

export const reset = (entries: Array<string | Object>, index: ?number, kind: ?string) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'reset',
    args: [entries, index, kind]
  }
})

export const setState = (state: Object | Function, n: ?(number | string), byIndex: ?boolean) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'setState',
    args: [state, n, byIndex]
  }
})

export const back = (state: ?(Object | Function)) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'back',
    args: [state]
  }
})

export const next = (state: ?(Object | Function)) => ({
  type: CALL_HISTORY,
  payload: {
    method: 'next',
    args: [state]
  }
})


// NOTE: see `src/utils/formatRoutes.js` for implemenation of corresponding pathlessRoute
