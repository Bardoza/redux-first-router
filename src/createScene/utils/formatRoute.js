// @flow
import { isNotFound, typeToScene, formatRoute } from '../../utils'

export default (r, type, routes, formatter) => {
  const route = formatRoute(r, type, routes, formatter)

  route.scene = typeToScene(type)

  // set default path for NOT_FOUND actions if necessary
  if (!route.path && isNotFound(type)) {
    route.path = route.scene
      ? `/${r.scene.toLowerCase()}/not-found`
      : '/not-found'
  }

  return route
}
