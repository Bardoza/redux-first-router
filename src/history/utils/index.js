export {
  supportsDom,
  supportsHistory,
  supportsSessionStorage
} from './supports'

export {
  addPopListener,
  removePopListener,
  isExtraneousPopstateEvent
} from './popListener'

export {
  saveHistory,
  restoreHistory,
  getHistoryState,
  getInitialHistoryState,
  getInitialN
} from './sessionStorage'
