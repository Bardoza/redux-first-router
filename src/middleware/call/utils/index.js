export { default as enhanceRoutes } from './enhanceRoutes'
export { default as onError } from './onError'
export { default as shouldCall } from './shouldCall'
export { default as createCache } from './createCache'
export { default as autoDispatch } from './autoDispatch'
