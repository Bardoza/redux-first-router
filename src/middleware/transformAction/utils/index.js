export { default as formatAction } from './formatAction'
export { default as replacePopAction, findNeighboringN } from './replacePopAction'
export { default as handleDoubleDispatch, isDoubleDispatch } from './doubleDispatch'
export { default as createNotFoundRoute } from './createNotFoundRoute'
export { default as nestAction, createActionRef, createStateRef } from './nestAction'

