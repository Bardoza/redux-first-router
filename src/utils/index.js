export { default as isHydrate } from './isHydrate'
export { default as isAction } from './isAction'
export { default as isNotFound } from './isNotFound'
export { default as isServer } from './isServer'
export { default as isRedirect } from './isRedirect'

export { default as actionToUrl } from './actionToUrl'
export { default as urlToAction } from './urlToAction'

export { default as doesRedirect } from './doesRedirect'
export { default as shouldTransition } from './shouldTransition'

export { default as matchUrl } from './matchUrl'
export { default as compileUrl } from './compileUrl'

export { default as formatRoutes, formatRoute } from './formatRoutes'
export { default as typeToScene } from './typeToScene'

export { default as noOp } from './noOp'

export { default as createSelector } from './createSelector'
export { default as createAction } from './createAction'

export { default as redirectShortcut } from './redirectShortcut'
export { default as callRoute } from './callRoute'
