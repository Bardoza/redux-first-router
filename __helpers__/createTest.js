import { applyMiddleware, createStore, combineReducers } from 'redux'
import { createRouter } from '../src'

export default async (...allArgs) => {
  const args = allArgs.filter(arg => typeof arg !== 'function')
  const callbacks = allArgs.filter(arg => typeof arg === 'function')
  const hasCallbacks = callbacks.length > 0

  const [testName, routesMap] = args
  let [,, options = {}, actions] = args

  if (Array.isArray(options)) {
    actions = options
    options = {}
  }

  if (options.skip) return

  options.browser = options.browser || false

  if (!actions) {
    actions = Object.keys(routesMap).filter(type => !/FIRST|REDIRECTED/.test(type))
  }

  const initialPath = typeof actions[0] === 'string' && actions[0].charAt(0) === '/'
    ? actions.shift()
    : '/first'

  const hasMultipleTests = actions.length > 1
  let num = 1

  if (hasMultipleTests) {
    describe(testName, () => {
      for (const action of actions) {
        if (Array.isArray(action)) {
          const name = action[0]
          const act = action[1]
          createTest(name, routesMap, initialPath, act, options, num++)
        }
        else {
          const name = JSON.stringify(action)
          createTest(name, routesMap, initialPath, action, options, num++)
        }
      }

      if (hasCallbacks) {
        callbacks.forEach((cb, index) => {
          const name = callbacks.length === 1 ? 'snipes' : 'snipes ' + (++index)
          createSnipes(name, routesMap, initialPath, options, cb)
        })
      }
    })
  }
  else {
    const isSnipesOnly = initialPath === '/first' && actions.length === 0

    if (hasCallbacks && isSnipesOnly) {
      callbacks.forEach((cb, index) => {
        const name = callbacks.length === 1 ? testName : 'snipes ' + (++index)
        createSnipes(name, routesMap, initialPath, options, cb)
      })
    }
    else if (hasCallbacks) {
      describe(testName, () => {
        const name = actions[0] ? JSON.stringify(actions[0]) : 'firstRoute - ' + initialPath

        createTest(name, routesMap, initialPath, actions[0], options, num)

        callbacks.forEach((cb, index) => {
          const name = callbacks.length === 1 ? 'snipes' : 'snipes ' + (++index)
          createSnipes(name, routesMap, initialPath, options, cb)
        })
      })
    }
    else if (Array.isArray(actions[0])) {
      const name = actions[0][0]
      const act = actions[0][1]

      createTest(name, routesMap, initialPath, act, options, num)
    }
    else {
      createTest(testName, routesMap, initialPath, actions[0], options, num)
    }
  }
}

const createTest = (testName, routesMap, initialPath, item, opts, num) => {
  test(testName, async () => {
    const {
      store,
      history,
      routes,
      options,
      firstRoute,
      initialState
    } = setupStore(routesMap, initialPath, opts)

    const firstAction = firstRoute(false)
    const res = await store.dispatch(firstAction)

    if (routesMap.FIRST || initialPath !== '/first') {
      const prefix = 'firstRoute - ' + initialPath + ' - ' + num
      snapChange(prefix, res, store, history, initialState)
    }

    if (typeof item === 'string' && item.charAt(0) === '/') {
      const url = item
      const res = await history.push(url)

      snapChange(num, res, store, history)
    }
    else if (item) {
      const action = typeof item === 'string' ? { type: item } : item
      const res = await store.dispatch(action)

      snapChange(num, res, store, history)
    }

    snapRoutes(num, routes)
    snapOptions(num, options)

    if (opts.log) console.log(store.getState().location)
  })
}

const createSnipes = (testName, routesMap, initialPath, opts, callback) => {
  test(testName, async () => {
    const {
      store,
      history,
      routes,
      options,
      firstRoute
    } = setupStore(routesMap, initialPath, opts)

    const firstResponse = opts.dispatchFirstRoute === false
      ? null
      : await store.dispatch(firstRoute(false))

    const pop = opts.browser && createPop(history)

    await callback({
      firstRoute,
      firstResponse,
      history,
      routes,
      options,
      store,
      pop,
      dispatch: store.dispatch,
      getState: store.getState,
      getLocation: () => store.getState().location,
      snapChange,
      snap: async (action, prefix = '') => {
        const res = await store.dispatch(action)

        prefix = prefix || JSON.stringify(action) || 'function'

        snapChange(prefix, res, store, history)
        snapRoutes(prefix, routes)
        snapOptions(prefix, options)

        return res
      },
      snapPop: async (direction, prefix = '') => {
        const res = await pop(direction)

        snapChange(prefix, res, store, history)
        snapRoutes(prefix, routes)
        snapOptions(prefix, options)

        return res
      }
    })

    if (opts.log) console.log(store.getState().location)
  })
}

export const setupStore = (routesMap, initialPath, opts) => {
  const routes = createRoutes(routesMap)
  const options = createOptions(opts)

  options.initialEntries = [initialPath]
  options.extra = options.extra || { arg: 'extra-arg' }

  const middlewareFunc = options.middlewareFunc
  delete options.middlewareFunc

  const title = (state, action = {}) => {
    return action.payload !== undefined
      ? action.type + ' - ' + JSON.stringify(action.payload)
      : action.type
  }

  const { middleware, reducer, firstRoute, history } = createRouter(
    routes,
    options,
    middlewareFunc
  )

  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware)
  const store = createStore(rootReducer, enhancer)

  const initialState = store.getState()

  return {
    store,
    firstRoute,
    initialState,
    routes,
    options,
    history
  }
}

const callbacks = [
  'beforeLeave',
  'beforeEnter',
  'onLeave',
  'onLeave',
  'onEnter',
  'thunk',
  'onComplete',
  'onError'
]

const createRoutes = (routesMap) => {
  const routes = {}

  for (const type in routesMap) {
    routes[type] = typeof routesMap[type] === 'object'
      ? { ...routesMap[type] }
      : routesMap[type]

    const route = routes[type]

    for (const cb of callbacks) {
      if (typeof route[cb] === 'function') {
        route[cb] = jest.fn(route[cb])
      }
    }
  }

  return {
    FIRST: {
      path: '/first'
    },
    NEVER_USED_PATHLESS: { // insures pathless routes can co-exist with regular routes
      thunk: jest.fn()
    },
    REDIRECTED: {
      path: '/redirected',
      beforeEnter: () => {
        return new Promise(res => setTimeout(res, 1))
      },
      onComplete: jest.fn(() => 'redirect_complete')
    },
    ...routes
  }
}

const createOptions = (opts = {}) => {
  const options = { ...opts }

  for (const cb of callbacks) {
    if (typeof opts[cb] === 'function') {
      options[cb] = jest.fn(opts[cb])
    }
  }

  return options
}

const snapRoutes = (prefix, routes) => {
  for (const type in routes) {
    const route = routes[type]
    snapCallbacks(prefix + ' - routes - ' + type, route)
  }
}

const snapOptions = (prefix, options) => {
  snapCallbacks(prefix + ' - options', options)
}

const snapCallbacks = (prefix, obj) => {
  expectBeforeLeave(prefix, obj)
  expectBeforeEnter(prefix, obj)
  expectOnLeave(prefix, obj)
  expectOnEnter(prefix, obj)
  expectThunk(prefix, obj)
  expectOnComplete(prefix, obj)
  expectOnError(prefix, obj)
}

const snapChange = (prefix, res, store, history, initialState) => {
  if (initialState) expectInitialState(prefix, initialState)

  expectState(prefix, store)
  expectResponse(prefix, res)
  expectEntries(prefix, history)
  expectIndex(prefix, history)
  expectTitle(prefix)
}

// all these expect functions are broken out separately so we can easily see the
// name of the expectation that failed in the trace displayed in the Wallaby web panel

const expectInitialState = (prefix, initialState) => {
  expect(initialState).toMatchSnapshot(prefix + ' - initial_state')
}

const expectState = (prefix, store) => {
  expect(store.getState()).toMatchSnapshot(prefix + ' - state')
}

const expectResponse = (prefix, res) => {
  expect(res).toMatchSnapshot(prefix + ' - response')
}

const expectEntries = (prefix, history) => {
  expect(history.entries).toMatchSnapshot(prefix + ' - history_entries')
}

const expectIndex = (prefix, history) => {
  expect(history.index).toMatchSnapshot(prefix + ' - history_index')
}

const expectTitle = (prefix) => {
  expect(document.title).toMatchSnapshot(prefix + ' - title')
}

const expectBeforeLeave = (prefix, obj) => {
  if (typeof obj.beforeLeave === 'function' && obj.beforeLeave.mock) {
    expect(obj.beforeLeave.mock.calls.length).toMatchSnapshot(prefix + ' - beforeLeave')
  }
}

const expectBeforeEnter = (prefix, obj) => {
  if (typeof obj.beforeEnter === 'function' && obj.beforeEnter.mock) {
    expect(obj.beforeEnter.mock.calls.length).toMatchSnapshot(prefix + ' - beforeEnter')
  }
}

const expectOnLeave = (prefix, obj) => {
  if (typeof obj.onLeave === 'function' && obj.onLeave.mock) {
    expect(obj.onLeave.mock.calls.length).toMatchSnapshot(prefix + ' - onLeave')
  }
}

const expectOnEnter = (prefix, obj) => {
  if (typeof obj.onEnter === 'function' && obj.onEnter.mock) {
    expect(obj.onEnter.mock.calls.length).toMatchSnapshot(prefix + ' - onEnter')
  }
}

const expectThunk = (prefix, obj) => {
  if (typeof obj.thunk === 'function' && obj.thunk.mock) {
    expect(obj.thunk.mock.calls.length).toMatchSnapshot(prefix + ' - thunk')
  }
}

const expectOnComplete = (prefix, obj) => {
  if (typeof obj.onComplete === 'function' && obj.onComplete.mock) {
    expect(obj.onComplete.mock.calls.length).toMatchSnapshot(prefix + ' - onComplete')
  }
}

const expectOnError = (prefix, obj) => {
  if (typeof obj.onError === 'function' && obj.onError.mock) {
    expect(obj.onError.mock.calls.length).toMatchSnapshot(prefix + ' - onError')
  }
}


// for testing pops in a "real" browser (as simulated by Jest)

const createPop = (history) => {
  return (direction = 'back') => {
    history.currentPop = null
    window.history[direction]()
    return awaitPop(history)
  }
}

const awaitPop = async (history, tries = 1) => {
  if (tries >= 10) {
    throw new Error(`awaitPop reached the maximum amount of tries (${tries})`)
  }

  await new Promise(res => setTimeout(res, 3))
  return history.currentPop || awaitPop(history, ++tries)
}
