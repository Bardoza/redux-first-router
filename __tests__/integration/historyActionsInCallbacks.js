import createTest, { resetBrowser } from '../../__helpers__/createTest'
import { get } from '../../src/history/utils'
import { push, replace } from '../../src/actions'

beforeEach(resetBrowser)

createTest('redirect before enter', {
  FIRST: '/',
  SECOND: {
    path: '/second',
    beforeEnter: async ({ dispatch }) => {
      await dispatch(push('/redirected'))
      // await dispatch(replace('/redirected'))
      // await dispatch({ type: 'REDIRECTED' })
    },
    thunk: async function({ dispatch }) {
      await dispatch(replace('/redirected'))
    }
  }
}, { testBrowser: true, onComplete: async ({ type, dispatch, routes }) => {
  console.log('GET!!', type, get(), window.location.pathname)

  if (type === 'REDIRECTED') {
    await new Promise(res => setTimeout(res, 50))
    // await dispatch(push('/not-found'))
    // await dispatch({ type: 'NOT_FOUND' })
  }
} })

// createTest('redirect after enter', {
//   FIRST: '/',
//   SECOND: {
//     path: '/second',
//     thunk: jest.fn(({ dispatch }) => {
//       return dispatch({ type: 'REDIRECTED' })
//     }),
//     onComplete: jest.fn()
//   }
// }, { testBrowser: true })

// createTest('redirect before enter (on firstRoute)', {
//   FIRST: {
//     path: '/',
//     beforeEnter: ({ dispatch }) => {
//       return dispatch({ type: 'REDIRECTED' })
//     },
//     thunk: function() {}
//   }
// }, { testBrowser: true })


// createTest('redirect after enter (on firstRoute)', {
//   FIRST: {
//     path: '/',
//     thunk: ({ dispatch }) => {
//       return dispatch({ type: 'REDIRECTED' })
//     },
//     onComplete: function() {}
//   }
// }, { testBrowser: true })
