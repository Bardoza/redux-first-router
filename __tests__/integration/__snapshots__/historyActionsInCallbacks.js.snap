// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redirect before enter: 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/",
      "scene": "",
      "search": "",
      "url": "/",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "FIRST",
  },
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/redirected",
      "scene": "",
      "search": "",
      "url": "/redirected",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "REDIRECTED",
  },
]
`;

exports[`redirect before enter: 1 - history_index 1`] = `1`;

exports[`redirect before enter: 1 - options - onComplete 1`] = `2`;

exports[`redirect before enter: 1 - response 1`] = `
Object {
  "payload": "redirect_complete",
  "type": "REDIRECTED_COMPLETE",
}
`;

exports[`redirect before enter: 1 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`redirect before enter: 1 - routes - REDIRECTED - onComplete 1`] = `1`;

exports[`redirect before enter: 1 - routes - SECOND - beforeEnter 1`] = `1`;

exports[`redirect before enter: 1 - routes - SECOND - thunk 1`] = `0`;

exports[`redirect before enter: 1 - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
    Array [
      "/redirected",
      Object {},
      "345678",
    ],
  ],
  "index": 1,
}
`;

exports[`redirect before enter: 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/redirected",
          "scene": "",
          "search": "",
          "url": "/redirected",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "REDIRECTED",
      },
    ],
    "from": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 1,
        "key": "345678",
        "pathname": "/second",
        "scene": "",
        "search": "",
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
    "hash": "",
    "index": 1,
    "key": "345678",
    "kind": "replace",
    "length": 2,
    "n": 1,
    "params": Object {},
    "pathname": "/redirected",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": 0,
        "key": "345678",
        "pathname": "/",
        "scene": "",
        "search": "",
        "url": "/",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "FIRST",
    },
    "query": Object {},
    "ready": true,
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 302,
    "type": "REDIRECTED",
    "universal": false,
    "url": "/redirected",
  },
  "title": "REDIRECTED_COMPLETE - \\"redirect_complete\\"",
}
`;

exports[`redirect before enter: 1 - title 1`] = `"REDIRECTED"`;

exports[`redirect before enter: 1 - windowLocation 1`] = `"/redirected"`;

exports[`redirect before enter: firstRoute - /first - 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "key": "345678",
      "pathname": "/",
      "scene": "",
      "search": "",
      "url": "/",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "FIRST",
  },
]
`;

exports[`redirect before enter: firstRoute - /first - 1 - history_index 1`] = `0`;

exports[`redirect before enter: firstRoute - /first - 1 - initial_state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "init",
    "length": 1,
    "n": 1,
    "params": Object {},
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "@@redux/INIT",
}
`;

exports[`redirect before enter: firstRoute - /first - 1 - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "index": 0,
    "key": "345678",
    "kind": "load",
    "length": 1,
    "n": 1,
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "FIRST",
}
`;

exports[`redirect before enter: firstRoute - /first - 1 - sessionStorage 1`] = `
Object {
  "entries": Array [
    Array [
      "/",
      Object {},
      "345678",
    ],
  ],
  "index": 0,
}
`;

exports[`redirect before enter: firstRoute - /first - 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "blocked": null,
    "direction": "forward",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "location": Object {
          "key": "345678",
          "pathname": "/",
          "scene": "",
          "search": "",
          "url": "/",
        },
        "params": Object {},
        "query": Object {},
        "state": Object {},
        "type": "FIRST",
      },
    ],
    "from": null,
    "hash": "",
    "index": 0,
    "key": "345678",
    "kind": "load",
    "length": 1,
    "n": 1,
    "params": Object {},
    "pathname": "/",
    "pop": false,
    "prev": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "index": -1,
        "key": "",
        "pathname": "",
        "scene": "",
        "search": "",
        "url": "",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "universal": false,
    "url": "/",
  },
  "title": "FIRST",
}
`;

exports[`redirect before enter: firstRoute - /first - 1 - title 1`] = `"FIRST"`;

exports[`redirect before enter: firstRoute - /first - 1 - windowLocation 1`] = `"/"`;
